### PR TITLE
Dt 44 refactor button component classes

### PR DESCRIPTION
**Pull request by juliodialpad**
## Description
We need to refactor our button component classes because they are used as Less mixins and will not be resolved unless the class name is completely defined in the Dialtone Less files

https://switchcomm.atlassian.net/browse/DT-44?atlOrigin=eyJpIjoiNGYyNzViYTc4Yjk4NDkyY2I0MDk0NTk5ZmVlZjBlZTMiLCJwIjoiaiJ9

## Pull Request Checklist

 - [x] Ensure all `gulp` scripts successfully compile.
 - [x] Update, remove, or extend all affected documentation.
 - [x] Provide a description what is being changed, extended, or removed.
 - [x] Link to any issue(s) this request is resolving.

### CODE REVIEW

File: lib/build/less/components/button.less [hunk 1]
```diff
@@ -8,147 +8,158 @@
...
+  pointer-events: none;
+}
+[disabled],
+.d-btn--disabled {
+  color: var(--black-400);
+  background-color: var(--black-100);
+  cursor: not-allowed;
+  border-color: transparent;
+  transition: none;
+}
```
Comment (ryandialpad) [inline lib/build/less/components/button.less]: This will apply styles to any element with `disabled`

Example:
```html
<input id="my-radio" type="radio" disabled />
<label for="my-radio">My radio</label>
```

The radio would have all of the styles listed above.

Your best bet for refactoring this would be to create an internal mixin which is used for the nested `[disabled]` selector and the disabled btn class.

File: lib/build/less/components/button.less [hunk 2]
```diff
@@ -8,147 +8,158 @@
...
+
+  .d-svg {
+    // no icon-size 12 exists so setting manually
+    width: 1.2rem !important;
+    height: 1.2rem !important;
+  }
+}
+.d-btn--circle.d-btn--xs {
+  padding: calc(var(--su8) - var(--su1));
 }
```
Comment (ryandialpad) [inline lib/build/less/components/button.less]: We should revert the changes to the circle class refactor as we only need the root classes to be un-nested so that they can be resolved. The nested selectors will work as intended.

Example:
```html
<button class="d-btn--circle d-btn--xs" />
```

```less
.my-custom-btn-class {
  .d-btn--circle;
  .d-btn--xs;
}
```

**Review by ryandialpad — CHANGES_REQUESTED** [review]
* Revert changes to circle btn size un-nesting
* Move `[disabled]` styles back in `d-btn` to prevent any element with `disabled` from being selected with those styles

**Review by juliodialpad — COMMENTED** [review]
Reverted the un-nesting and moved [disabled] states

File: lib/build/less/components/button.less [hunk 3]
```diff
@@ -8,147 +8,149 @@
...
-        font-size: var(--fs24);
-        line-height: var(--lh8);
-
-        .d-svg {
-            .d-svg--size24();
-        }
-    }
+//  $$  DISABLED
+//  ----------------------------------------------------------------------------
+.d-btn {
```
Comment (ryandialpad) [inline lib/build/less/components/button.less]: `d-btn` is defined twice, could we please move these into the existing `d-btn` class above?

File: lib/build/less/components/button.less [hunk 4]
```diff
@@ -190,191 +192,205 @@
...
+}
+.d-btn__icon--left {
+  margin-left: var(--sun2);
+  margin-right: var(--su4);
+}
+.d-btn__icon--right {
+  order: 1;
+  margin-left: var(--su4);
+  margin-right: var(--sun2);
 }
```
Comment (ryandialpad) [inline lib/build/less/components/button.less]:
```suggestion
.d-btn__icon {
  display: flex;
  margin-top: var(--sun4);
  margin-bottom: var(--sun4);
  width: 1em;
  height: 1em;
}

.d-btn__icon--left {
  margin-left: var(--sun2);
  margin-right: var(--su4);
}

.d-btn__icon--right {
  order: 1;
  margin-left: var(--su4);
  margin-right: var(--sun2);
}
```

Ensure that you have padding between class definitions

Comment (ryandialpad) [reply lib/build/less/components/button.less]: Same goes with circle button size selectors

File: lib/build/less/components/button.less (hunk 3 above)
Comment (ryandialpad) [reply lib/build/less/components/button.less]: `d-btn--disabled` should also be defined at root level so it can be resolved

Example
```less
.d-btn {
  ...
  &[disabled] {
    color: var(--black-400);
    background-color: var(--black-100);
    cursor: not-allowed;
    border-color: transparent;
    transition: none;
  }
  ...
}

.d-btn--disabled {
  pointer-events: none;
  color: var(--black-400);
  background-color: var(--black-100);
  cursor: not-allowed;
  border-color: transparent;
  transition: none;
}
```

You can build on this example if you want to make the following styles a little more re-usable

```less
color: var(--black-400);
background-color: var(--black-100);
cursor: not-allowed;
border-color: transparent;
transition: none;
```

File: lib/build/less/components/button.less (hunk 4 above)
Comment (ryandialpad) [reply lib/build/less/components/button.less]: and the button icon classes

**Review by ryandialpad — CHANGES_REQUESTED** [review]
See comments

File: lib/build/less/components/button.less [hunk 5]
```diff
@@ -131,23 +143,23 @@
...
+    &.d-btn__icon--left {
+        margin-left: var(--sun2);
+        margin-right: var(--su4);
+    }
 
+    &.d-btn__icon--right {
+        order: 1;
+        margin-left: var(--su4);
+        margin-right: var(--sun2);
+    }
```
Comment (ryandialpad) [inline lib/build/less/components/button.less]: These still need to be un-nested to be resolved as mixins

**Review by ryandialpad — APPROVED** [review]
w/ one remaining change

See comments
